modules/gdscript/gdscript_analyzer.cpp

```diff
@@ -1929,49 +1885,50 @@ void GDScriptAnalyzer::reduce_array(GDScriptParser::ArrayNode *p_array) {
 
 // When an array literal is stored (or passed as function argument) to a typed context, we then assume the array is typed.
 // This function determines which type is that (if any).
-void GDScriptAnalyzer::update_array_literal_element_type(const GDScriptParser::DataType &p_base_type, GDScriptParser::ArrayNode *p_array_literal) {
-	GDScriptParser::DataType array_type = p_array_literal->get_datatype();
-	if (p_array_literal->elements.size() == 0) {
-		// Empty array literal, just make the same type as the storage.
-		array_type.set_container_element_type(p_base_type.get_container_element_type());
-	} else {
-		// Check if elements match.
-		bool all_same_type = true;
-		bool all_have_type = true;
-
-		GDScriptParser::DataType element_type;
-		for (int i = 0; i < p_array_literal->elements.size(); i++) {
-			if (i == 0) {
-				element_type = p_array_literal->elements[0]->get_datatype();
-			} else {
-				GDScriptParser::DataType this_element_type = p_array_literal->elements[i]->get_datatype();
-				if (this_element_type.has_no_type()) {
-					all_same_type = false;
-					all_have_type = false;
-					break;
-				} else if (element_type != this_element_type) {
-					if (!is_type_compatible(element_type, this_element_type, false)) {
-						if (is_type_compatible(this_element_type, element_type, false)) {
-							// This element is a super-type to the previous type, so we use the super-type.
-							element_type = this_element_type;
-						} else {
-							// It's incompatible.
-							all_same_type = false;
-							break;
-						}
-					}
-				}
-			}
+void GDScriptAnalyzer::update_array_literal_element_type(GDScriptParser::ArrayNode *p_array, const GDScriptParser::DataType *p_base_type) {
+	GDScriptParser::DataType array_type = p_array->get_datatype();
+
+	GDScriptParser::DataType base_element_type;
+	bool has_base_element_type = p_base_type != nullptr && p_base_type->has_container_element_type();
+	if (has_base_element_type) {
+		base_element_type = p_base_type->get_container_element_type();
+		base_element_type.is_constant = false;
```

Comment:
Would it be appropriate to use `type_from_metatype` here?

Reply:
Have not encountered this method call in any places of type resolution. Seems like it does not do much except removing info about original enum. You saying that `[Enum.A, Enum.B]` should be inferred as `Array[int]` and not `Array[Enum]`?

Reply:
I'm not really referring to enums - just in general to "metatypes".

It's a weird method. I tend to think of a `metatype` as the `DataType` referring to the class itself, whereas if it isn't a metatype, then it refers to an instance of that class. Similarly, are you referring to the `enum` itself, or to a variable with a type of that enum.

So, if the base element type is `class A`, which is a `metatype` and incidentally also `constant`, then you can use `type_from_metatype` to say that elements are instances of the class `A`. Otherwise you're saying elements of the array are the class itself, not instances of the class.

Though to be fair, it almost seems like `is_metatype` and `is_constant` are used interchangeably :)

Reply:
I think more specifically, if you use `resolve_datatype` or `reducer_identifier` on the `[A]` of `var my_array : Array [A]`, then the datatype will be a `metatype`, but you want elements to _not_ be `metatype`. Does that make sense at all?

Reply:
Seems like it does not apply here then. If you have `[A, B]` where `A` and `B` are classes (not an instances) then it should be typed as an array of classes and not array of instances of those classes.

Reply:
I'll trust you on it! Don't have capacity to dig very deep :)

My main concern was 
```GDScript

var arr : Array[A] = [A.new(), A.new()]
var my_var := arr[0]
```

I just want to make sure that the type of `my_var` is not a metatype. If it isn't, then we're all good!

Reply:
Ah, yes, disregard previous comment. By that point if there is a specified datatype it will be called `resolve_datatype` on it. And [those lines](https://github.com/godotengine/godot/blob/6318320c04f348eee1df7a5dae3bad540ef1fdcc/modules/gdscript/gdscript_analyzer.cpp#L492-L494) make sure that the element type in not metatype.

Reply:
Perfect! :D